### PR TITLE
Negative coin error fix

### PR DESCRIPTION
**Pull request by arhamchordia**
## 1. Overview

Adding the condition which avoids negative coin error.

## 2. Implementation details

- Adds a condition to return an error If the undelegated amount is more than already staked.
- Adds the condition which avoids negative coin error.

## 3. How to test/use

- Run the test function added to make sure every case is covered.

## 4. Checklist

- [ ] Does the Readme need to be updated?

## 6. Future Work (optional)

Need a well defined strategy implemented from scratch for whichever the latest version of the product will be or the current strategy can be generalised a little to make it better.

### CODE REVIEW

**Review by Anmol1696 — APPROVED** [review]
LGTM

File: x/lscosmos/keeper/delegation_strategy.go [hunk 1]
```diff
@@ -168,8 +168,14 @@ func distributeCoinsAmongstValSet(ws types.WeightedAddressAmounts, coin sdk.Coin
 			valAddrAmts = append(valAddrAmts, types.ValAddressAmount{ValidatorAddr: w.Address, Amount: coin})
 			return valAddrAmts, sdk.NewInt64Coin(coin.Denom, 0)
 		}
-		valAddrAmts = append(valAddrAmts, types.ValAddressAmount{ValidatorAddr: w.Address, Amount: w.Coin()})
-		coin = coin.SubAmount(w.Amount)
+
+		tempAmount := w.Amount
+		if w.Amount.IsNegative() {
+			tempAmount = sdk.ZeroInt()
```
Comment (puneet2019) [inline x/lscosmos/keeper/delegation_strategy.go:174]: i think negatives can be returned early..

Comment (arhamchordia) [reply x/lscosmos/keeper/delegation_strategy.go:185]: I don't see the need in the test cases I have added.

Comment (puneet2019) [reply x/lscosmos/keeper/delegation_strategy.go:185]: not need, just small optimisation

File: x/lscosmos/keeper/delegation_strategy.go [hunk 2]
```diff
@@ -91,23 +92,33 @@ func (k Keeper) UndelegateMsgs(ctx sdk.Context, amount sdk.Int, denom string, de
...
 	return DivideUndelegateAmountIntoValidatorSet(currDiffDistribution, amount)
 }
 
 // GetIdealCurrentDelegations returns ideal amount of delegations to validators on host chain
-func GetIdealCurrentDelegations(valList types.AllowListedValidators, delegationState types.DelegationState, amt sdk.Coin, reverse bool) types.WeightedAddressAmounts {
+func GetIdealCurrentDelegations(valList types.AllowListedValidators, delegationState types.DelegationState, amt sdk.Coin, reverse bool) (types.WeightedAddressAmounts, error) {
 	totalDelegations := delegationState.TotalDelegations(amt.Denom)
 
+	if reverse && totalDelegations.IsLT(amt) {
+		return nil, fmt.Errorf("undelegation amount greater than already staked, staked:  %s, undelegate : %s", totalDelegations, amt)
```
Comment (puneet2019) [inline x/lscosmos/keeper/delegation_strategy.go:119]: use sdkerrors

Comment (arhamchordia) [reply x/lscosmos/keeper/delegation_strategy.go:119]: Yes

**Review by puneet2019 — APPROVED** [review]

**Review by Anmol1696 — APPROVED** [review]
LGTM

**Pull request merged**